File: fibgo/fib.go
package main

import (
    "fmt"
    "time"
)

// fib is a pure, recursive function to calculate the Nth Fibonacci number.
// We use the standard 'int' type, which is usually 64-bit on modern systems,
// sufficient to hold the result for N=40.
func fib(n int) int {
    if n <= 1 {
        return n
    }
    // Recursive calls
    return fib(n-1) + fib(n-2)
}

func main() {
    // --- Configuration ---
    // Choose a value for N, consistent with the other benchmarks.
    const n = 40

    // --- Benchmarking ---
    // Start timing using time.Now().
    start := time.Now()

    // Calculate the Fibonacci number.
    result := fib(n)

    // Calculate the duration since 'start'.
    elapsed := time.Since(start)

    // Convert duration to milliseconds for comparison.
    elapsed_ms := float64(elapsed.Microseconds()) / 1000.0

    // --- Output ---
    fmt.Println("--- Go Recursive Fibonacci Benchmark ---")
    fmt.Println("N:", n)
    fmt.Println("Result:", result)
    fmt.Printf("Time to calculate: %.3f ms\n", elapsed_ms)
    fmt.Println("--------------------------------------")
}

File: fibzig/fib.zig
const std = @import("std");

// fib is a pure, recursive function to calculate the Nth Fibonacci number.
// We use u64 (unsigned 64-bit integer) to ensure the result is correctly typed.
fn fib(n: u64) u64 {
    if (n <= 1) {
        return n;
    }
    // Recursive calls
    return fib(n - 1) + fib(n - 2);
}

// The main function returns !void, indicating it might throw an error (e.g., during I/O).
pub fn main() !void {
    // --- Configuration ---
    // Choose a value for N, consistent with the other benchmarks.
    const n: u64 = 40;

    // Get a writer for standard output, handling potential writer errors with `!`.
    //const stdout = std.fs.File.stdout();

    // --- Benchmarking ---
    // Start timing using the high-resolution nanoTimestamp.
    const start_nanos = std.time.nanoTimestamp();

    // Calculate the Fibonacci number.
    const result = fib(n);

    // Calculate the duration.
    const end_nanos = std.time.nanoTimestamp();
    const elapsed_nanos: f64 = @floatFromInt(end_nanos - start_nanos);

    // Convert duration to milliseconds (1,000,000 nanoseconds per millisecond).
    const elapsed_ms: f64 = elapsed_nanos / 1_000_000.0;

    // --- Output ---
    const print = @import("std").debug.print;
    print("--- Zig Recursive Fibonacci Benchmark ---\n", .{});
    print("N: {}\n", .{n});
    print("Result: {}\n", .{result});
    // Use the {d:.3} format specifier to print the float with 3 decimal places.
    print("Time to calculate: {d:.3} ms\n", .{elapsed_ms});
    print("--------------------------------------\n", .{});
}

